### DISCUSSION TITLE
Implementation of "#embed"

### DISCUSSION THREAD

**Discussion by lauler1** [hsutter/cppfront#784] [Q&A] [answered]
Is there any plan to add [Jean-Heyd Meneide’s proposal for #embed](https://wg21.link/p1967) in Cppfront before C++26?
It was mentioned briefly by Herb Sutter [here](https://herbsutter.com/).

**Comment by JohelEGP** [comment]
I don't think Cppfront can implement that.
Not with the performance expectations of `#embed`.

**Reply by JohelEGP** [reply]
But maybe program-defined metafunctions can subsume a subset of its use cases
(if program-defined metafunctions will be "just functions").
Then you can just use your usual I/O libraries.
The built-in one for `@print` uses `std::cout`,
and there's plans for (<https://github.com/hsutter/cppfront/discussions/650#discussioncomment-7149419>):
> writing a metafunction that generates an output file at compile time

**Reply by AbhinavK00** [reply]
This makes me curious, can there be a metafunction in cpp1 which takes in a function and then writes out the cpp2 equivalent of it in another file (or maybe generate a copy of the file but with that one function in cpp2)? Could be possible but not feasible I guess, is that right.?

**Reply by JohelEGP** [reply]
Do you mean like a tool to convert Cpp1 to Cpp2?
That doesn't need to be implemented with a metafunction.

**Reply by AbhinavK00** [reply]
Not a full fledged tool, I'm just wondering about the limits of capabilities of metafunctions.

**Reply by JohelEGP** [reply]
That's a question for Herb.
I'm hoping metafunctions are just functions.

**Reply by hsutter** [reply]
If I understand the question correctly, metafunctions are just functions. You could `#include <reflect.h>` and write them in Cpp1 syntax... that's how most of the ones we have now began life.

I'm not planning anything here about a tool to convert Cpp1 to Cpp2. I anticipated that if we got that far, it would be done by extending a Cpp1 compiler front-end.

**Comment by hsutter** [comment, marked as answer]
I haven't thought about anything here related to `#embed`.
